.ci/scripts/generate_cc_changelog/generate.py

```diff
@@ -0,0 +1,346 @@
+        for pull in commit_pulls:
+            ret.append(pull)
+    return ret
+
+
+def gather_shas(working_tree):
+    """
+    Discover a set of SHAs in the history between the first and the last,
+    inclusively
+
+    Parameters
+    ----------
+    working_tree : str, required
+        The git repo to examine. Must pass in root of repo.
+
+    Returns
+    -------
+    list
+        A list of commits. Each commit is a string.
+    """
+    ret = list()
+    repo = git.Repo(working_tree)
+    latest_tag = repo.tags.pop()
+    for sha in repo.iter_commits(rev="{tag}..HEAD".format(tag=latest_tag)):
+        ret.append(str(sha))
+    return ret
+
+
+def sha_of_last_tag(working_tree):
+    """
+    Return the SHA of the previous tag
+
+    Parameters
+    ----------
+    working_tree : str, required
+        The git repo to examine. Must pass in root of repo.
+
+    Returns
+    -------
+    git.Commit
+        The most recent commit. Must wrap in str() to eval as a
+        a string.
+    """
+    repo = git.Repo(working_tree)
+    assert not repo.bare
+    latest_tag = repo.tags.pop()
+    return latest_tag.commit
+
+
+def extract_releases(changelog_path):
+    """
+    Take an existing changelog and return a list of changelog entries
+
+    Parameters
+    ----------
+    changelog_path : str, required
+        The path of the CHANGELOG.asciidoc file to examine
+
+    Returns
+    -------
+    list
+        A list of changelog entries. Each element in the list is a list
+        which contains all the plain-text lines from that entry, complete
+        with newlines.
+    """
+    release_blocks = list()
+    fh_ = open(changelog_path, "r")
+    release_block = list()
+    release_section_found = False
+    for line in fh_:
+        if re.match("\[\[release-notes-", line):  # noqa: W605
+            # We found a release note block, if we had a release block already,
+            # go ahead and record it.
+            if release_block:
+                release_blocks.append(release_block)
+            # Now clear out the release block.
+            release_block = list()
+            # Add this first line of the block to the block container
+            release_block.append(line)
+            # We are in the release section so go ahead and set the flag to
+            # continue collection
+            release_section_found = True
+        elif release_section_found:
+            release_block.append(line)
+
+    return release_blocks
+
+
+def is_changelog_worthy(pr):
+    """
+    Take a PR title and return a boolean indicating whether or not the PR
+    should be included in a changelog
+
+    Parameters
+    ----------
+    pr : github.PullRequest
+
+    Returns
+    -------
+    bool
+        True if the pull request is one that we consider as being appropriate
+        for the changelog. See the implementation for details on how this
+        selection process operates.
+    """
+    if re.match("fix(\(.*\)){0,1}:|\
```

Comment:
I see there are more types defined in https://github.com/conventional-changelog/commitlint , is this intended to be the only ones to be searched?

Reply:
This is drawn from the list of previous changelog entries for this project. I had to speculate about what is (and isn't) included. Perhaps @watson could shed some more light on what kinds of changelogs entries they wish to include and exclude?

Reply:
We tend to only put user-facing commits in the changelog because we had the philosophy that if we put all commits there, we might as well just link to the commit log instead.

User-facing commits are always `feat`, `fix`, and `perf` and sometimes `chore` and `refactor`. But never `test`, `ci`, or `docs`. Previously we've had the luxury of being able to cherry-pick which of the `chore` and `refactor` commits to include, but we can't do that if we automate the process. There's a ton of `chore` commits in most releases that are just noise - e.g. when dev-dependency packages are upgraded. It would be sad to include those. So I propose the following:

1. We always include `feat`, `fix`, and `perf`
1. We always include any commit that contains a breaking change, no matter the type.

Rule 2 would mean that `refactor!: drop support for Node 6` would be included in the changelog even though `refactor` normally isn't. I think this would solve our problems.

Speaking of breaking changes: Recently Conventional Commits released v1.0.0 of their spec, which added the ability to denote breaking changes with an exclamation mark (`!`) before the colon, e.g. `refactor!: drop support for Node 6`. So there are two ways to mark a commit as containing a breaking change:

1. Using `!` in the first line (see [rule 1](https://www.conventionalcommits.org/en/v1.0.0/#specification))
1. Using `BREAKING CHANGE: ...` in the footer (see [rule 12](https://www.conventionalcommits.org/en/v1.0.0/#specification))

Reply:
Hi @watson 

Thanks for this excellent write-up. I have modified the detection script to handle breaking changes by searching for breaking changes using both the rules that you mention.

One question, though. Let's say that (hypothetically), you had a PR which had to separate commits which would qualify for the CHANGELOG. Would you prefer to have (a) a single entry for the CHANGELOG which give the title of the PR in question or would you prefer to have (b) two separate entries, one for each qualifying commit, both of point to the same PR?

The script currently uses the behavior described in (a) but I can modify it if you think the second is preferable. Let me know what you think.

Reply:
Good question... How will you currently know which PR title to pick in case of (a)? Do you rely on the PR ID to be in the commit message? If we use "Rebase and merge" to merge a PR so that the individual commits are retained, the PR ID will _not_ be added to the commit message. This only happens when you use "Squash and merge" - which is one reason why I don't like "Rebase and merge". So are you sure that your current implementation works in that case?

Reply:
In the case of (a) the PR title is used and it is fetched by querying the GitHub API. In the case of (b), we examine the git history directly instead of querying GitHub.

Reply:
Ah cool - I didn't know GitHub retained that information 👍 

This project has rarely used the old merge nor the rebase-and-merge in the past. Normally we've always used squash-and-merge. But I just went back to see an examples of one of the few times we did use rebase-and-merge:

PR: #1636
PR title: Fix CI
PR description: _\<blank\>_
Commits:
- test(pg): fix pg TAV tests
- ci(tav): fix @hapi/hapi TAV tests
- ci(tav): fix got TAV tests
- test: fix APM Server cert test to work with 13.9.0+
- fix: support promisifying setTimeout and friends …
- ci(jenkins): ensure postgres container doesn't need password
- fix(cassandra): improve support for cassandra-driver v4.4.0+ …
- test: fix APM Server schema tests

In this case option (b) would have been best as we didn't want the title "Fix CI", but would have liked the following to be added to the changelog:

- fix: support promisifying setTimeout and friends
- fix(cassandra): improve support for cassandra-driver v4.4.0+

### Conclusion

In this light I think option (b) would be best.

Reply:
Btw, some of the commits on purpose have a body with a detailed explanation that can't fit in the commit message headline. I think those bodies should go into the release notes as well if we can (previously we have not done that, but I think it would be worth investigating).

For example, the two above "fix" commits actually both had bodies, eg:

```
fix: support promisifying setTimeout and friends

When using `util.promisify` on functions where the callback is not the
last argument, a custom promisify function is normally provided on a
symbol on the main function. This is also the case with `setTimeout` and
`setImmediate`.

Our patch meant that `setTimeout` and friends would no longer have this
symbol on them and as such would not be promisified correctly.

This commit looks for the presence of such a promisifier and copies it
over from the original function to the wrapper function.

This commit also applies the same logic for patching functions where the
callback isn't the first argument. Currently we don't know of a case
where there is a similar issue with these kind of functions, but we
might as well protect those functions from this sort of issue while
we're at it.
```

Reply:
Sounds good! Thank you for the clear analysis. I'll make the necessary changes.

Reply:
OK, I have pushed a change which allows us to use either the GitHub strategy (a) or the git-based strategy (b) and defaults to git-based.

As a part of this, I also ensured that the git-strategy pushes the entire body of the commit message into the changelog [as requested](https://github.com/elastic/apm-agent-nodejs/pull/1714#discussion_r414356442). Therefore, I think the requirements of this request are now satisfied and I'm going to go ahead and resolve this thread.